tools/asset-sync/assets.ps1

```
@@ -0,0 +1,521 @@
+#>
+Function Is-AssetsRepo-Initialized {
+    param(
+        [PSCustomObject] $Config
+    )
+
+    $result = $false
+    $assetRepoLocation = Resolve-AssetRepo-Location -Config $Config
+
+    try {
+        Push-Location $assetRepoLocation
+        $gitLocation = Join-Path $assetRepoLocation ".git"
+
+        if (Test-Path $gitLocation){
+            $result = $true
+        }
+        else{
+            $result = $false
+        }
+    }
+    catch {
+        Write-Error $_
+        $result = $false
+    }
+    finally {
+        Pop-Location
+    }
+
+    return $result
+}
+
+
+<#
+.SYNOPSIS
+Given a configuration, determine which paths must be added to the sparse checkout of the assets repo.
+
+.PARAMETER Config
+A PSCustomObject that contains an auto-parsed assets.json content.
+
+#>
+Function Resolve-CheckoutPaths {
+    param(
+        [Parameter(Mandatory=$true)]
+        [PSCustomObject] $Config
+    )
+    $assetsJsonPath = $Config.AssetsJsonRelativeLocation
+
+    $assetsJsonFolder = Split-Path $Config.AssetsJsonRelativeLocation
+
+    if(!$assetsJsonFolder -or $assetsJsonFolder -in @(".", "./", ".\"))
+    {
+        return $Config.AssetsRepoPrefixPath.Replace("`\", "/")
+    }
+    else {
+        $result = (Join-Path $Config.AssetsRepoPrefixPath $assetsJsonFolder).Replace("`\", "/")
+        return $result
+    }
+}
+
+<#
+.SYNOPSIS
+Given a configuration, determine the _current_ target path.
+
+.DESCRIPTION 
+Determines the presence of a branch on the git repo. If the relevant auto/<service> branch does not exist, we should use main.
+#>
+Function Resolve-TargetBranch {
+    param(
+        [Parameter(Mandatory=$true)]
+        [PSCustomObject] $Config
+    )
+    $assetRepo = Resolve-AssetRepo-Location -Config $Config
+    $branch = "main"
+    try {
+        Push-Location $assetRepo
+        
+        $latestCommit = git rev-parse "origin/$($Config.AssetsRepoBranch)"
```

If the branch doesn't exist, you get exception text on stdErr:

```
fatal: ambiguous argument 'origin/auto/test': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
origin/auto/test
```

To disambiguate and silence the error text:
```
$latestCommit = git rev-parse "origin/$($Config.AssetsRepoBranch)" -- 2>$null
```

---

It's necessary to disambiguate because someone could have the local relative path `origin/main`. If they do, `git rev-parse origin/main` returns `origin/main`

---

It's necessary to disambiguate. If the user has the local relative path `origin/auto/something` and not the ref, `git rev-parse origin/auto/something` will return the local relative path and not a commit id or 128 exit code.

---

Done. Thanks patrick!